dipy/segment/tissue.py

```diff
@@ -0,0 +1,122 @@
+import numpy as np
+from dipy.sims.voxel import add_noise
+from dipy.segment.mrf import (ConstantObservationModel,
+                              IteratedConditionalModes)
+
+
+class TissueClassifierHMRF(object):
+    r"""
+    This class contains the methods for tissue classification using the Markov
+    Random Fields modeling approach
+    """
+
+    def __init__(self, save_history=False, verbose=True):
+
+        self.save_history = save_history
+        self.segmentations = []
+        self.pves = []
+        self.energies = []
+        self.energies_sum = []
+        self.verbose = verbose
+
+        pass
+
+    def classify(self, image, nclasses, beta, tolerance):
+        r"""
+        This method uses the Maximum a posteriori - Markov Random Field
+        approach for segmentation by using the Iterative Conditional Modes and
+        Expectation Maximization to estimate the parameters.
+
+        Parameters
+        ----------
+        image : ndarray,
+                3D structural image.
+        nclasses : int,
+                    number of desired classes.
+        beta : float,
+                smoothing parameter, the higher this number the smoother the
+                output will be.
+        max_iter : float,
+                    number of desired iterations.
+
+        Returns
+        -------
+        initial_segmentation : ndarray,
+                                3D segmented image with all tissue types
+                                specified in nclasses.
+        final_segmentation : ndarray,
+                                3D final refined segmentation containing all
+                                tissue types.
+        PVE : ndarray,
+                3D probability map of each tissue type.
+        """
+
+        nclasses = nclasses + 1  # One extra class for the background
+        energy_sum = []
+
+        com = ConstantObservationModel()
+        icm = IteratedConditionalModes()
+
+        if image.max() > 1:
+            image = np.interp(image, [0, image.max()], [0.0, 1.0])
+
+        mu, sigma = com.initialize_param_uniform(image, nclasses)
+
+#        mu = np.array([0.0, 0.03, 0.5, 0.8])
+#        sigma = np.array([0.01, 0.01, 0.15, 0.15])
+        sigmasq = sigma ** 2
+
+        neglogl = com.negloglikelihood(image, mu, sigmasq, nclasses)
+        seg_init = icm.initialize_maximum_likelihood(neglogl)
+
+        mu, sigma = com.seg_stats(image, seg_init, nclasses)
+        sigmasq = sigma ** 2
+
+        zero = np.zeros_like(image) + 0.001
+        zero_noise = add_noise(zero, 10000, 1, noise_type='gaussian')
+        image_gauss = np.where(image == 0, zero_noise, image)
+
+        final_segmentation = np.empty_like(image)
+        initial_segmentation = seg_init.copy()
+
+        for i in range(100):
```

Comment:
We should not hard-code the number of iterations. I think users will expect this to be an input parameter.